crypto/s2n_hash.c

```diff
@@ -55,8 +56,18 @@ S2N_RESULT s2n_hash_algorithms_init()
      * Additionally, the old style methods do not support property query strings
      * to guide which provider to fetch from. This is important for FIPS, where
      * the default query string of "fips=yes" will need to be overridden for
-     * legacy algorithms.
+     * legacy algorithms. TODO: are they actually still in use in FIPS mode?
+     *
      */
+    if (!OSSL_PROVIDER_available(NULL, "default")) {
+      /* System is configured without the default provider, but we are
+       * about to fetch MD5 and SHA1 from it below, note note sure if
+       * this should be conditional and opportunistic, as most
+       * policies no longer use SHA1, and even SHA224 is weak these
+       * days. Similar approach is taken in python to fetch MD5 when
+       * usedforsecurity=False */
+         OSSL_PROVIDER_load(NULL, "default");
+    }
```

Comment:
Historically, s2n-tls does not configure Openssl for customers. Any changes we make will affect other libraries / application use of Openssl. If we load "default" here for NULL library context, we're loading "default" globally with no indication of how that will affect other code that depends on Openssl. We require that the application configure Openssl as needed / desired.

Reply:
Ack - but the other documentation in s2n-tls README explicitly tells end user to load globally and system-wide for all applications and the system, including s2n-tls but also any other system apps.

My ideal preference is to actually not load the default provider at all - and instead not crash and silently not allow MD5 / SHA1 policies etc, and be more oportunistic. Right now s2n-tls crashes if MD5 is not there, and default provider is not loaded.

Would it be acceptable to make MD5, SHA1, MD5-SHA1 optional when default provider has not already been loaded by the user in the global config and/or within the app context that uses s2n-tls?

Reply:
yeah, but like effectively all FIPS is now TLS1.2+ only in OpenSSL context => the non-extended-master-secret TLSv1.2 has been retired, which needs at least openssl 1.1.1 which has tlsv1.2 and tlsv1.3 support. And if one otherwise compiles openssl in seclevel 2 by default, and sets min prototocl to tlsv1.2 it is a strong indicator the user does not want these legacy algorithms.

Which is quite contratry to what s2n-tls stands for - to be a universal terminator, with own implementation and configs of arbitrary tls feature sets.

Let me ponder things more around here.